ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,204 @@
+
+import static org.opensearch.ml.engine.algorithms.remote.ConnectorUtils.processOutput;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.apache.logging.log4j.util.Strings;
+import org.opensearch.OpenSearchStatusException;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.ml.common.connector.Connector;
+import org.opensearch.ml.common.output.model.ModelTensors;
+import org.opensearch.script.ScriptService;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+
+@Log4j2
+public class MLSdkAsyncHttpResponseHandler implements SdkAsyncHttpResponseHandler {
+    @Getter
+    private Integer statusCode;
+    @Getter
+    private final StringBuilder responseBody = new StringBuilder();
+
+    private WrappedCountDownLatch countDownLatch;
+
+    private ActionListener<List<ModelTensors>> actionListener;
+
+    private Map<String, String> parameters;
+
+    private Map<Integer, ModelTensors> tensorOutputs;
+
+    private Connector connector;
+
+    private ScriptService scriptService;
+
+    private final List<String> errorMsg = new ArrayList<>();
+
+    public MLSdkAsyncHttpResponseHandler(
+        WrappedCountDownLatch countDownLatch,
+        ActionListener<List<ModelTensors>> actionListener,
+        Map<String, String> parameters,
+        Map<Integer, ModelTensors> tensorOutputs,
+        Connector connector,
+        ScriptService scriptService
+    ) {
+        this.countDownLatch = countDownLatch;
+        this.actionListener = actionListener;
+        this.parameters = parameters;
+        this.tensorOutputs = tensorOutputs;
+        this.connector = connector;
+        this.scriptService = scriptService;
+    }
+
+    @Override
+    public void onHeaders(SdkHttpResponse response) {
+        SdkHttpFullResponse sdkResponse = (SdkHttpFullResponse) response;
+        log.debug("received response headers: " + sdkResponse.headers());
+        this.statusCode = sdkResponse.statusCode();
+    }
+
+    @Override
+    public void onStream(Publisher<ByteBuffer> stream) {
+        stream.subscribe(new MLResponseSubscriber());
+    }
+
+    @Override
+    public void onError(Throwable error) {
+        log.error(error.getMessage(), error);
+        actionListener
+            .onFailure(
+                new OpenSearchStatusException(
+                    "Error on communication with remote model: " + error.getMessage(),
+                    RestStatus.INTERNAL_SERVER_ERROR
```

Comment:
Why always return this internal server error?

Reply:
Can we translate the status code to this RestStatus

Reply:
Done